test/htmlrenderer.cpp

```diff
@@ -1204,3 +1204,144 @@ TEST_CASE("When rendeing text, HtmlRenderer strips leading whitespace",
+	REQUIRE(links.size() == 0);
+}
+
+TEST_CASE("<audio> results in a placeholder and a link for each valid source",
+	"[HtmlRenderer]")
+{
+	HtmlRenderer r;
+
+	const std::string input =
+		"<audio src='http://example.com/audio.oga'></audio>"
+		"<audio>"
+		"	<source src='http://example.com/audio2.mp3'>"
+		"	<source src='http://example.com/audio2.m4a'>"
+		"</audio>";
+	std::vector<std::pair<LineType, std::string>> lines;
+	std::vector<LinkPair> links;
+
+	REQUIRE_NOTHROW(r.render(input, lines, links, url));
+	REQUIRE(lines.size() == 6);
+	REQUIRE(lines[0] == p(LineType::wrappable, "[audio 1 (link #1)]"
+			"[audio 2 (link #2)]"
+			"[audio 2 (link #3)]"));
+	REQUIRE(lines[1] == p(LineType::wrappable, ""));
+	REQUIRE(lines[2] == p(LineType::wrappable, "Links: "));
+	REQUIRE(lines[3] ==
+		p(LineType::softwrappable,
+			"[1]: http://example.com/audio.oga (audio)"));
+	REQUIRE(lines[4] ==
+		p(LineType::softwrappable,
+			"[2]: http://example.com/audio2.mp3 (audio)"));
+	REQUIRE(lines[5] ==
+		p(LineType::softwrappable,
+			"[3]: http://example.com/audio2.m4a (audio)"));
+	REQUIRE(links.size() == 3);
+	REQUIRE(links[0].first == "http://example.com/audio.oga");
+	REQUIRE(links[0].second == LinkType::AUDIO);
+	REQUIRE(links[1].first == "http://example.com/audio2.mp3");
+	REQUIRE(links[1].second == LinkType::AUDIO);
+	REQUIRE(links[2].first == "http://example.com/audio2.m4a");
+	REQUIRE(links[2].second == LinkType::AUDIO);
+}
+
+TEST_CASE("<audio>s without valid sources are ignored", "[HtmlRenderer]")
+{
+	HtmlRenderer r;
+
+	const std::string input = "<audio></audio>"
+		"<audio><source><source></audio>";
+	std::vector<std::pair<LineType, std::string>> lines;
+	std::vector<LinkPair> links;
+
+	REQUIRE_NOTHROW(r.render(input, lines, links, url));
+	REQUIRE(lines.size() == 0);
+	REQUIRE(links.size() == 0);
+}
+
+TEST_CASE("Unclosed <video> and <audio> tags are closed upon encounter with a "
+	"new media element", "[HtmlRenderer]")
+{
+	HtmlRenderer r;
+
+	const std::string input =
+		"<video src='http://example.com/video.avi'>"
+		"	This is fallback text for `the video` element"
+		"<video>"
+		"	<source src='http://example.com/video2.avi'>"
+		"This maybe isn't fallback text, but the spec says that"
+		" anything before the closing tag is transparent content"
+		"<audio>"
+		"	<source src='http://example.com/audio.oga'>"
+		"	<source src='http://example.com/audio.m4a'>"
+		"	This text should also be interpreted as fallback"
+		"<audio src='http://example.com/audio2.mp3'>"
+		"	This is additional fallback text"
+		"<audio></audio>"
+		"Here comes the text!";
+
+	std::vector<std::pair<LineType, std::string>> lines;
+	std::vector<LinkPair> links;
+
+	REQUIRE_NOTHROW(r.render(input, lines, links, url));
+	REQUIRE(lines.size() == 8);
+	REQUIRE(lines[1] == p(LineType::wrappable, ""));
+	REQUIRE(lines[2] == p(LineType::wrappable, "Links: "));
+	REQUIRE(links.size() == 5);
+	REQUIRE(lines[0] == p(LineType::wrappable, "[video 1 (link #1)]"
```

Comment:
It's a bit confusing that this line is at the end—please move it up, before `lines[1]` check.

Also, please check the contents of the `links`, to ensure the URLs were assigned the correct numbers.

Reply:
Okay!